# Lab 1 6,000xp


## Part 1 - Fun with Crickets

<img src="http://bugwoodcloud.org/images/384x256/5387688.jpg" width="400" />

A long, long time ago, in the summer of 1898, Ernst Athearn Bessey and his brother Carl decided to observe tree crickets in Lincoln, Nebraska. (Ernst was 21 at the time and just finished his MA degree) They were interested in the relationship between the speed of cricket chirps and outdoor temperature. The thing that made this a bit easier was that they found "that each cricket remained in the same tree for days at a time."



<img src="http://www.cybertruffle.org.uk/pics/0001727a.jpg" width= "300" />


We are going to examine the data they collected.

First let's import the numpy library



In [3]:
import numpy as np

Now we will read the data file from the web and place it in an np.array

In [8]:
## read in the cricket file
cricket_data = np.genfromtxt('http://zacharski.org/files/crickets.csv', delimiter=',')

In [10]:
print(cricket_data)

[[ 20.          88.59999847]
 [ 16.          71.59999847]
 [ 19.79999924  93.30000305]
 [ 18.39999962  84.30000305]
 [ 17.10000038  80.59999847]
 [ 15.5         75.19999695]
 [ 14.69999981  69.69999695]
 [ 17.10000038  82.        ]
 [ 15.39999962  69.40000153]
 [ 16.20000076  83.30000305]
 [ 15.          79.59999847]
 [ 17.20000076  82.59999847]
 [ 16.          80.59999847]
 [ 17.          83.50000125]
 [ 14.39999962  76.30000305]
 [ 17.1         81.50055115]
 [ 13.1         68.36003334]
 [ 13.65        70.        ]
 [ 14.2         72.00011123]
 [ 18.67        86.67711113]
 [ 19.1226667   88.21000242]
 [ 19.01        84.05999912]
 [ 17.61        83.18999999]
 [ 16.125       77.30100011]
 [ 13.          67.26666667]
 [ 12.          62.72740003]
 [ 12.49        66.33333333]
 [ 14.111       77.81221677]
 [ 17.6         85.        ]]


That first column is the number of chirps and the second column is the temperature.


<h2 style="color:red">Task 1: Descriptive Statistics: 800xp</h2>
What are the mean and median for both the cricket chirps and the temperature? I would like your report to look like

    Cricket chirps
        mean: 16.1237471341
      median: 16.125

    Temperature
        mean: 78.3116698272
      median: 80.59999847

In [1]:
# Your code here

<h2 style="color:red">Task 2: Correlation: 800xp</h2>

Using `corrcoef` I want to know whether there is a correlation between the number of chirps and the temperature. Can you compute the correlation?

In [ ]:
# your code here

<h2 style="color:red">Cricket Hacker Challenge Part 1: 1000xp</h2>

A year before the Bessey brothers counted cricket chirps, Amos Dolbear (shown below)did the same. Dolbear was a physics professor at Tufts University and invented the telephone some 11 years before Bell.

<img src="https://upload.wikimedia.org/wikipedia/commons/9/94/Charles_Edwin_Bessey.jpg" width="300"/>


When Dolbear was 60 he came up with what is now known as Dolbear's Law can predict temeperature in Fahrenheight from cricket chirps per minute ($N_{60}$):
        

### $$T_F = 50 + \frac{N_{60} - 40}{4}$$

Can you create a one-column numpy array with the predicted values of Fahrenheit based on the Cricket Chirp column of cricket_data?  As a hint, if you have a numpy array:

In [21]:
old = np.array([10, 20, 30, 40, 50])

and if my formula were 

$$new = old + 100$$

I can create the new array using:

In [22]:
new = old + 100

In [23]:
new


array([110, 120, 130, 140, 150])

In [25]:
# Your predicted code here:








<h2 style="color:red">Cricket Hacker Challenge Part 2: 1000xp</h2>
    
This is going to require some thinking.  That's why it is called a hacker challenge.

We now have the predicted values based on Dolbear's Law and we have the actual values. I would like to know the average error. Here is what I mean by that. 





Suppose I think I am pretty good at guessing people's weight. Here are my estimates for 2 people and their actual weights (which I got after I guessed):

My Guess |  Actual Weight |
:--: | :--:
200 |  150 
125 |  175 


Now I would like to know on average how much I was off. (so on average I was off x pounds per guess). If you can do this example with pencil and paper you are well on your way to cricket success.

numpy's absolute value function may be helpful here:


In [27]:
x = np.array([10, -10, -15, 20, 5])
np.abs(x)

array([10, 10, 15, 20,  5])

In [1]:
# Do your work here


# Part 2: Smoking and Birth Weight

First I want to expand on what was covered in the Datacamp course. Suppose I have a numpy array that represents this table:


Gender | Weight
 :--: | :--: |
 Female | 93
 Female | 113.8
 Female | 137
 Male | 183.8
 Female | 110
 Male | 152
 
 So I create a numpy array:
 

In [13]:
students = np.array([['Female', 'Female', 'Female', 'Male', 'Female', 'Male',], [93, 138, 137, 183.8, 110, 152]])
print(students)


[['Female' 'Female' 'Female' 'Male' 'Female' 'Male']
 ['93' '138' '137' '183.8' '110' '152']]


We can extract just the weights of the women by:

In [12]:
females = students[1][students[0] == 'Female']
print(females)

['93' '1138' '137' '110']


Now on to our data...

## Birth Rate Data

The birth rate data looks like this:

ID | LOW | AGE | LWT | RACE | SMOKE | PTL | HT  | UI | FTV | BWT
:--: | :--: | :--: | :--: | :--: | :--: | :--: | :--: | :--: | :--: | :--: 
85 | 0 | 19 | 182 | 2 | 0 | 0 | 0 | 1 | 0 | 2523
86 | 0 | 33 | 155 | 3 | 0 | 0 | 0 | 0 | 3 | 2551
87 | 0 | 20 | 105 | 1 | 1 | 0 | 0 | 0 | 1 | 2557
88 | 0 | 21 | 108 | 1 | 1 | 0 | 0 | 1 | 2 | 2594
89 | 0 | 18 | 107 | 1 | 1 | 0 | 0 | 1 | 0 | 2600
91 | 0 | 21 | 124 | 3 | 0 | 0 | 0 | 0 | 0 | 2622


First, let's load the data into numpy:


In [18]:

birth_data = np.transpose(np.genfromtxt('http://zacharski.org/files/courses/data101/birthweight.csv', delimiter=',', dtype=int))
print(birth_data)

[[  85   86   87 ...,   82   83   84]
 [   0    0    0 ...,    1    1    1]
 [  19   33   20 ...,   23   17   21]
 ..., 
 [   1    0    0 ...,    0    0    0]
 [   0    3    1 ...,    0    0    3]
 [2523 2551 2557 ..., 2495 2495 2495]]


In printing the numpy array things look a little flipped. So the first column (the ID column) of the original data is represented by the first row of the numpy array. Oddly enough this is exactly how we want it. 


Here is what the columns mean. 

* **ID** the first column is simply the id number of the mother
* **BWT** The last column is the birth weight in grams.
* **LOW** A zero in this column indicates the birth weight is not considered low a one indicates it is
* **AGE** The age of the mother
* **LWT** 
* **RACE** 1: white, 2: black, 3: other
* **SMOKE** whether or not the person smokes
* **PTL** number of pre-mature labors
* **HT** Hypertension
* **UI** uterine irratability
* **FTV**  GP visits


<h2 style="color:red">Task 3: Descriptive Statistics on the Entire Dataset: 1000xp</h2>
I would like to know the minimum, average (mean), and maximum birth weights of the dataset in a format that looks like:

    BIRTH WEIGHTS
    minimum:   00 grams
    mean:      00 grams
    maximum:   00 grams
    std dev    00 grams
    
There is a <span style="color:red">200xp bonus</span>, if instead your print out looks like:


    BIRTH WEIGHTS
    minimum:   0 pounds 0 ounces
    mean:      0 pounds 0 ounces
    maximum:   0 pounds 0 ounces
    std dev.   0 pounds 0 ounces
 

<h2 style="color:red">Task 4: Smokers vs Non-Smokers 1000xp</h2>

Now I would like to see those statistics for smokers vs. non-smokers

    SMOKERS BIRTH WEIGHTS
    minimum:   00 grams
    mean:      00 grams
    maximum:   00 grams    
    std dev    00 grams
    
    NONSMOKERS BIRTH WEIGHTS
    minimum:   00 grams
    mean:      00 grams
    maximum:   00 grams   
    std dev.   00 grams
    
What do you think? Are the differences significant enough for you to think there is a difference between smokers and nonsmokers? (we will learn more formal ways of defining significance later in the course)

# Congratulations! 
You just finished a fairly hard unit on Numpy and looked at some real datasets.